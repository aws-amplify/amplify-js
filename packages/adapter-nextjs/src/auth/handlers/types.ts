// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { OAuthConfig } from '@aws-amplify/core';
import { NextApiRequest, NextApiResponse } from 'next';
import { CookieStorage } from 'aws-amplify/adapter-core';

import { CreateAuthRoutesHandlersInput } from '../types';

interface AuthApiRequestHandlerInputBase {
	oAuthConfig: OAuthConfig;
	origin: string;
	userPoolClientId: string;
	setCookieOptions: CookieStorage.SetCookieOptions;
}

// handleSignInRequest
interface HandleSignInSignUpRequestInputBase
	extends AuthApiRequestHandlerInputBase {
	customState?: string;
	type: 'signIn' | 'signUp';
}
interface HandleSignInSignUpRequestInput
	extends HandleSignInSignUpRequestInputBase {
	request: Request;
}
interface HandleSignInSigUpRequestForPagesRouterInput
	extends HandleSignInSignUpRequestInputBase {
	request: NextApiRequest;
	response: NextApiResponse;
}
export type HandleSignInSignUpRequest = (
	input: HandleSignInSignUpRequestInput,
) => Response;
export type HandleSignInSignUpRequestForPagesRouter = (
	input: HandleSignInSigUpRequestForPagesRouterInput,
) => void;

// handleSignInCallbackRequest
interface HandleSignInCallbackRequestInput
	extends AuthApiRequestHandlerInputBase {
	request: Request;
	handlerInput: CreateAuthRoutesHandlersInput;
}
interface HandleSignInCallbackRequestForPagesRouterInput
	extends AuthApiRequestHandlerInputBase {
	request: NextApiRequest;
	response: NextApiResponse;
	handlerInput: CreateAuthRoutesHandlersInput;
}
export type HandleSignInCallbackRequest = (
	input: HandleSignInCallbackRequestInput,
) => Promise<Response>;
export type HandleSignInCallbackRequestForPagesRouter = (
	input: HandleSignInCallbackRequestForPagesRouterInput,
) => Promise<void>;

// handleSignOutRequest
type handleSignOutRequestInput = AuthApiRequestHandlerInputBase;
interface handleSignOutRequestForPagesRouterInput
	extends AuthApiRequestHandlerInputBase {
	response: NextApiResponse;
}
export type HandleSignOutRequest = (
	input: handleSignOutRequestInput,
) => Response;
export type HandleSignOutRequestForPagesRouter = (
	input: handleSignOutRequestForPagesRouterInput,
) => void;

// handleSignOutCallbackRequest
interface HandleSignOutCallbackRequestInput
	extends Omit<AuthApiRequestHandlerInputBase, 'origin'> {
	request: Request;
	handlerInput: CreateAuthRoutesHandlersInput;
}
interface HandleSignOutCallbackRequestForPagesHandlerInput
	extends Omit<AuthApiRequestHandlerInputBase, 'origin'> {
	request: NextApiRequest;
	response: NextApiResponse;
	handlerInput: CreateAuthRoutesHandlersInput;
}
export type HandleSignOutCallbackRequest = (
	input: HandleSignOutCallbackRequestInput,
) => Promise<Response>;
export type HandleSignOutCallbackRequestForPagesRouter = (
	input: HandleSignOutCallbackRequestForPagesHandlerInput,
) => Promise<void>;
