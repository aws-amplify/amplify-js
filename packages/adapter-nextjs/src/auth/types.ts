// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ResourcesConfig } from 'aws-amplify';
import { NextRequest } from 'next/server';
import { NextApiRequest, NextApiResponse } from 'next';
import { OAuthConfig } from '@aws-amplify/core';
import { CookieStorage } from 'aws-amplify/adapter-core';

import { NextServer } from '../types';

export type SupportedRoutePaths =
	| 'sign-in'
	| 'sign-up'
	| 'sign-in-callback'
	| 'sign-out'
	| 'sign-out-callback';

/**
 * The Auth API routes can be handled by the {@link AuthRouteHandlers}.
 */
export interface AuthRouteHandlerParams {
	slug: string | string[] | undefined;
}

export interface AuthRoutesHandlerContext {
	params: AuthRouteHandlerParams;
}

/**
 * The handler function for handling the GET requests sent to the Auth API routes.
 * Only `GET` method gets handled, otherwise it rejects.
 *
 * @param request - `request` can be he following:
 *   1. a `NextRequest` when the handler is used in the App Router of Next.js
 *   2. a `NextApiRequest` when the handler is used in the Pages Router of Next.js
 * @param contextOrRequest - `contextOrRequest` can be the following:
 *   1. a {@link AuthRoutesHandlerContext} when the handler is used in the App
 *      Router of Next.js
 *   2. a `NextApiResponse` when the handler is used in the Pages Router of Next.js
 *
 * @returns a `Promise` of `Response` when used in the App Router of Next.js, or
 *   returns `undefined` when used in the Pages Router of Next.js.
 */
export type AuthRouteHandlers = (
	request: NextRequest | NextApiRequest,
	contextOrRequest: AuthRoutesHandlerContext | NextApiResponse,
) => // Next.js API route handler is required to be an async function.
Promise<Response | undefined>;

export interface CreateAuthRoutesHandlersInput {
	/**
	 * The customer's custom state added as the redirect URL query parameter
	 * for preventing CSRF attacks
	 */
	customState?: string;
	/**
	 * The app route redirect to when a sign-in flow completes. Defaults o the
	 * root of the app if not provided.
	 */
	redirectOnSignInComplete?: string;
	/**
	 * The app route redirect to when a sign-out flow completes. Defaults o the
	 * root of the app if not provided.
	 */
	redirectOnSignOutComplete?: string;
}

export type CreateAuthRouteHandlers = (
	input?: CreateAuthRoutesHandlersInput,
) => AuthRouteHandlers;

export interface CreateAuthRouteHandlersFactoryInput {
	config: ResourcesConfig;
	runtimeOptions: NextServer.CreateServerRunnerRuntimeOptions | undefined;
}

export type CreateOAuthRouteHandlersFactory = (
	input: CreateAuthRouteHandlersFactoryInput,
) => CreateAuthRouteHandlers;

interface HandleAuthApiRouteRequestInputBase {
	handlerInput: CreateAuthRoutesHandlersInput;
	oAuthConfig: OAuthConfig;
	setCookieOptions: CookieStorage.SetCookieOptions;
	origin: string;
}

interface HandleAuthApiRouteRequestForAppRouterInput
	extends HandleAuthApiRouteRequestInputBase {
	request: NextRequest;
	handlerContext: AuthRoutesHandlerContext;
}

interface HandleAuthApiRouteRequestForPagesRouterInput
	extends HandleAuthApiRouteRequestInputBase {
	request: NextApiRequest;
	response: NextApiResponse;
}

export type HandleAuthApiRouteRequestForAppRouter = (
	input: HandleAuthApiRouteRequestForAppRouterInput,
) => Response;

export type HandleAuthApiRouteRequestForPagesRouter = (
	input: HandleAuthApiRouteRequestForPagesRouterInput,
) => void;
