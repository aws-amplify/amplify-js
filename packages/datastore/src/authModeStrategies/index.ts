// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export { defaultAuthStrategy } from './defaultAuthStrategy';
export { multiAuthStrategy } from './multiAuthStrategy';
