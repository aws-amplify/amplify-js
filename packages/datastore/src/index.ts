// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
	USER,
	isModelConstructor,
	isNonModelConstructor,
	traverseModel,
	validatePredicate,
} from './util';

export {
	DataStore,
	DataStoreClass,
	initSchema,
	ModelInstanceCreator,
	// eslint-disable-next-line import/export
	AsyncCollection,
	AsyncItem,
} from './datastore/datastore';

export {
	Predicates,
	ModelPredicateCreator,
	ModelSortPredicateCreator,
} from './predicates';
export { Adapter as StorageAdapter } from './storage/adapter';

export { NAMESPACES } from './util';

export const utils = {
	USER,
	traverseModel,
	validatePredicate,
	isNonModelConstructor,
	isModelConstructor,
};

// eslint-disable-next-line import/export
export * from './types';
