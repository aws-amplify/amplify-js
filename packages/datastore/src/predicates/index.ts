// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
	ModelPredicate,
	PersistentModel,
	PredicatesGroup,
	ProducerModelPredicate,
	SchemaModel,
} from '../types';
import { extractPrimaryKeyFieldNames, extractPrimaryKeyValues } from '../util';

export { ModelSortPredicateCreator } from './sort';

const predicatesAllSet = new WeakSet<ProducerModelPredicate<any>>();

export function isPredicatesAll(
	predicate: any,
): predicate is typeof PredicateAll {
	return predicatesAllSet.has(predicate);
}

/**
 * The valid logical grouping keys for a predicate group.
 */
const groupKeys = new Set(['and', 'or', 'not']);

/**
 * Determines whether an object is a GraphQL style predicate "group", which must be an
 * object containing a single "group key", which then contains the child condition(s).
 *
 * E.g.,
 *
 * ```
 * { and: [ ... ] }
 * { not: { ... } }
 * ```
 *
 * @param o The object to test.
 */
const isGroup = o => {
	const keys = [...Object.keys(o)];

	return keys.length === 1 && groupKeys.has(keys[0]);
};

/**
 * Determines whether an object specifies no conditions and should match everything,
 * as would be the case with `Predicates.ALL`.
 *
 * @param o The object to test.
 */
const isEmpty = o => {
	return !Array.isArray(o) && Object.keys(o).length === 0;
};

/**
 * The valid comparison operators that can be used as keys in a predicate comparison object.
 */
export const comparisonKeys = new Set([
	'eq',
	'ne',
	'gt',
	'lt',
	'ge',
	'le',
	'contains',
	'notContains',
	'beginsWith',
	'between',
]);

/**
 * Determines whether an object is a GraphQL style predicate comparison node, which must
 * be an object containing a single "comparison operator" key, which then contains the
 * operand or operands to compare against.
 *
 * @param o The object to test.
 */
const isComparison = o => {
	const keys = [...Object.keys(o)];

	return !Array.isArray(o) && keys.length === 1 && comparisonKeys.has(keys[0]);
};

/**
 * A light check to determine whether an object is a valid GraphQL Condition AST.
 *
 * @param o The object to test.
 */
const isValid = o => {
	if (Array.isArray(o)) {
		return o.every(v => isValid(v));
	} else {
		return Object.keys(o).length <= 1;
	}
};

// This symbol is not used at runtime, only its type (unique symbol)
export const PredicateAll = Symbol('A predicate that matches all records');

export class Predicates {
	public static get ALL(): typeof PredicateAll {
		// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
		const predicate = <ProducerModelPredicate<any>>(c => c);

		predicatesAllSet.add(predicate);

		return predicate as unknown as typeof PredicateAll;
	}
}

export class ModelPredicateCreator {
	/**
	 * Map of storage predicates (key objects) to storage predicate AST's.
	 */
	private static predicateGroupsMap = new WeakMap<
		ModelPredicate<any>,
		PredicatesGroup<any>
	>();

	/**
	 * Determines whether the given storage predicate (lookup key) is a predicate
	 * key that DataStore recognizes.
	 *
	 * @param predicate The storage predicate (lookup key) to test.
	 */
	static isValidPredicate<T extends PersistentModel>(
		predicate: any,
	): predicate is ModelPredicate<T> {
		return ModelPredicateCreator.predicateGroupsMap.has(predicate);
	}

	/**
	 * Looks for the storage predicate AST that corresponds to a given storage
	 * predicate key.
	 *
	 * The key must have been created internally by a DataStore utility
	 * method, such as `ModelPredicate.createFromAST()`.
	 *
	 * @param predicate The predicate reference to look up.
	 * @param throwOnInvalid Whether to throw an exception if the predicate
	 * isn't a valid DataStore predicate.
	 */
	static getPredicates<T extends PersistentModel>(
		predicate: ModelPredicate<T>,
		throwOnInvalid = true,
	) {
		if (throwOnInvalid && !ModelPredicateCreator.isValidPredicate(predicate)) {
			throw new Error('The predicate is not valid');
		}

		return ModelPredicateCreator.predicateGroupsMap.get(predicate);
	}

	/**
	 * using the PK values from the given `model` (which can be a partial of T
	 * Creates a predicate that matches an instance described by `modelDefinition`
	 * that contains only PK field values.)
	 *
	 * @param modelDefinition The model definition to create a predicate for.
	 * @param model The model instance to extract value equalities from.
	 */
	static createForPk<T extends PersistentModel>(
		modelDefinition: SchemaModel,
		model: T,
	) {
		const keyFields = extractPrimaryKeyFieldNames(modelDefinition);
		const keyValues = extractPrimaryKeyValues(model, keyFields);

		const predicate = this.createFromAST<T>(modelDefinition, {
			and: keyFields.map((field, idx) => {
				const operand = keyValues[idx];

				return { [field]: { eq: operand } };
			}),
		});

		return predicate;
	}

	/**
	 * Searches a `Model` table for records matching the given equalities object.
	 *
	 * This only matches against fields given in the equalities object. No other
	 * fields are tested by the predicate.
	 *
	 * @param modelDefinition The model we need a predicate for.
	 * @param flatEqualities An object holding field equalities to search for.
	 */
	static createFromFlatEqualities<T extends PersistentModel>(
		modelDefinition: SchemaModel,
		flatEqualities: Record<string, any>,
	) {
		const ast = {
			and: Object.entries(flatEqualities).map(([k, v]) => ({ [k]: { eq: v } })),
		};

		return this.createFromAST<T>(modelDefinition, ast);
	}

	/**
	 * Accepts a GraphQL style filter predicate tree and transforms it into an
	 * AST that can be used for a storage adapter predicate. Example input:
	 *
	 * ```js
	 * {
	 * 	and: [
	 * 		{ name: { eq: "Bob Jones" } },
	 * 		{ age: { between: [32, 64] } },
	 * 		{ not: {
	 * 			or: [
	 * 				{ favoriteFood: { eq: 'pizza' } },
	 * 				{ favoriteFood: { eq: 'tacos' } },
	 * 			]
	 * 		}}
	 * 	]
	 * }
	 * ```
	 *
	 * @param gql GraphQL style filter node.
	 */
	static transformGraphQLFilterNodeToPredicateAST(gql: any) {
		if (!isValid(gql)) {
			throw new Error(
				'Invalid GraphQL Condition or subtree: ' + JSON.stringify(gql),
			);
		}

		if (isEmpty(gql)) {
			return {
				type: 'and',
				predicates: [],
			};
		} else if (isGroup(gql)) {
			const groupkey = Object.keys(gql)[0];
			const children = this.transformGraphQLFilterNodeToPredicateAST(
				gql[groupkey],
			);

			return {
				type: groupkey,
				predicates: Array.isArray(children) ? children : [children],
			};
		} else if (isComparison(gql)) {
			const operatorKey = Object.keys(gql)[0];

			return {
				operator: operatorKey,
				operand: gql[operatorKey],
			};
		} else {
			if (Array.isArray(gql)) {
				return gql.map(o => this.transformGraphQLFilterNodeToPredicateAST(o));
			} else {
				const fieldKey = Object.keys(gql)[0];

				return {
					field: fieldKey,
					...this.transformGraphQLFilterNodeToPredicateAST(gql[fieldKey]),
				};
			}
		}
	}

	/**
	 * Accepts a GraphQL style filter predicate tree and transforms it into a predicate
	 * that storage adapters understand. Example input:
	 *
	 * ```js
	 * {
	 * 	and: [
	 * 		{ name: { eq: "Bob Jones" } },
	 * 		{ age: { between: [32, 64] } },
	 * 		{ not: {
	 * 			or: [
	 * 				{ favoriteFood: { eq: 'pizza' } },
	 * 				{ favoriteFood: { eq: 'tacos' } },
	 * 			]
	 * 		}}
	 * 	]
	 * }
	 * ```
	 *
	 * @param modelDefinition The model that the AST/predicate must be compatible with.
	 * @param ast The graphQL style AST that should specify conditions for `modelDefinition`.
	 */
	static createFromAST<T extends PersistentModel>(
		modelDefinition: SchemaModel,
		ast: any,
	): ModelPredicate<T> {
		const key = {} as ModelPredicate<T>;

		ModelPredicateCreator.predicateGroupsMap.set(
			key,
			this.transformGraphQLFilterNodeToPredicateAST(ast),
		);

		return key;
	}
}
