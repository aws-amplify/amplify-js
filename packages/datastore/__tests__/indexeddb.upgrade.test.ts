import Dexie from 'dexie';
import 'dexie-export-import';
import 'fake-indexeddb/auto';
import * as idb from 'idb';
import {
	initSchema,
	ModelInit,
	MutableModel,
	PersistentModelConstructor,
	DataStore,
} from '../src/index';
import { Schema } from '../src/types';

const DB_VERSION = 3;

let db: idb.IDBPDatabase;
const { indexedDB } = require('fake-indexeddb');
const IDBKeyRange = require('fake-indexeddb/lib/FDBKeyRange');
Dexie.dependencies.indexedDB = indexedDB;
Dexie.dependencies.IDBKeyRange = IDBKeyRange;

/* 
	This schema has only one of the models that exists in the v1 schema we're upgrading from
	below './v1schema.data.json'

	Previously, this would cause an error in the IDB upgrade process
	This test validates that the db upgrade works as expected
*/
const schema: Schema = {
	models: {
		Blog: {
			syncable: true,
			name: 'Blog',
			pluralName: 'Blogs',
			fields: {
				id: {
					name: 'id',
					isArray: false,
					type: 'ID',
					isRequired: true,
					attributes: [],
				},
			},
			attributes: [
				{
					type: 'model',
					properties: {},
				},
			],
		},
	},
	enums: {},
	nonModels: {},
	version: 'a66372d29356c40e7cd29e41527cead7',
	codegenVersion: '3.2.0',
};

declare class BlogModel {
	readonly id: string;
	readonly name: string;
	constructor(init: ModelInit<BlogModel>);
	static copyOf(
		source: BlogModel,
		mutator: (draft: MutableModel<BlogModel>) => MutableModel<BlogModel> | void,
	): BlogModel;
}

const { Blog } = initSchema(schema) as {
	Blog: PersistentModelConstructor<BlogModel>;
};

describe('DB versions migration with destructive schema change', () => {
	beforeEach(async () => {
		await DataStore.clear();
	});

	/* 
		Note: ensure fake-idb/Dexie data isolation if adding additional tests
	*/
	test(`Migration from v1 to v${DB_VERSION}`, async () => {
		const v1Data = require('./v1schema.data.json');

		const blob = new Blob([JSON.stringify(v1Data)], {
			type: 'application/json',
		} as any);

		// Import V1
		(await Dexie.import(blob)).close();

		// Migrate to latest
		await DataStore.start();

		const blogRes = await DataStore.query(Blog);
		const expectedCount = v1Data.data.tables.find(
			t => t.name === 'user_Blog',
		).rowCount;

		expect(blogRes.length).toEqual(expectedCount);

		// Open latest
		db = await idb.openDB('amplify-datastore', DB_VERSION);

		// 3 internal stores + 1 user store, user_Blog
		expect(db.objectStoreNames.length).toBe(4);

		for (const storeName of db.objectStoreNames) {
			expect(db.transaction(storeName).store.indexNames).toContain('byPk');
		}

		const dexie = await new Dexie('amplify-datastore').open();
		const exportedBlob = await dexie.export();

		function readBlob(blob: Blob): Promise<string> {
			return new Promise((resolve, reject) => {
				const reader = new FileReader();
				reader.onabort = ev => reject(new Error('file read aborted'));
				reader.onerror = ev => reject((ev.target as any).error);
				reader.onload = ev => resolve((ev.target as any).result);
				reader.readAsText(blob);
			});
		}

		const exportedJSON = await readBlob(exportedBlob);
		const exported = JSON.parse(exportedJSON);

		for (const { schema } of exported.data.tables) {
			expect(schema.split(',')).toContain('&[id]');
		}

		expect(exported).toMatchSnapshot(`v${DB_VERSION}-destructive-schema`);
	});
});
