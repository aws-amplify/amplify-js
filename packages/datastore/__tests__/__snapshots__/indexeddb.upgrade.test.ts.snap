// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DB versions migration with destructive schema change Migration from v1 to v3: v3-destructive-schema 1`] = `
{
  "data": {
    "data": [
      {
        "inbound": false,
        "rows": [
          [
            1,
            {
              "id": "094a722a-4f00-4723-928a-536aa04b3470",
              "key": "schemaVersion",
              "value": ""a66372d29356c40e7cd29e41527cead7"",
            },
          ],
        ],
        "tableName": "datastore_Setting",
      },
      {
        "inbound": false,
        "rows": [],
        "tableName": "sync_ModelMetadata",
      },
      {
        "inbound": false,
        "rows": [],
        "tableName": "sync_MutationEvent",
      },
      {
        "inbound": false,
        "rows": [
          {
            "$": [
              1,
              {
                "_deleted": null,
                "_lastChangedAt": null,
                "_version": null,
                "blogOwnerId": "1a232e28-14ed-43e7-b39c-beba80e62a1d",
                "id": "7649de35-1602-4a60-8d3e-4aecae73ebc6",
                "name": "blog2",
              },
            ],
            "$types": {
              "$": {
                "1._deleted": "undef",
                "1._lastChangedAt": "undef",
                "1._version": "undef",
              },
            },
          },
          {
            "$": [
              2,
              {
                "_deleted": null,
                "_lastChangedAt": null,
                "_version": null,
                "id": "844724fe-373c-4505-b202-2852580058e3",
                "name": "B1",
                "owner": {
                  "_deleted": null,
                  "_lastChangedAt": null,
                  "_version": null,
                  "id": "d3b11bac-699a-4ff5-8626-10cb6bda634a",
                  "name": "O1",
                },
              },
            ],
            "$types": {
              "$": {
                "1._deleted": "undef",
                "1._lastChangedAt": "undef",
                "1._version": "undef",
                "1.owner._deleted": "undef",
                "1.owner._lastChangedAt": "undef",
                "1.owner._version": "undef",
              },
            },
          },
          {
            "$": [
              3,
              {
                "_deleted": null,
                "_lastChangedAt": null,
                "_version": null,
                "id": "91e67f1e-aba4-497d-a2c1-f2d3133468c0",
                "name": "Avatar: Last Airbender",
                "owner": {
                  "_deleted": null,
                  "_lastChangedAt": null,
                  "_version": null,
                  "id": "210bd12c-8ca3-4c3c-b67c-911d84a875b8",
                  "name": "Owner 1",
                },
              },
            ],
            "$types": {
              "$": {
                "1._deleted": "undef",
                "1._lastChangedAt": "undef",
                "1._version": "undef",
                "1.owner._deleted": "undef",
                "1.owner._lastChangedAt": "undef",
                "1.owner._version": "undef",
              },
            },
          },
          {
            "$": [
              4,
              {
                "_deleted": null,
                "_lastChangedAt": null,
                "_version": null,
                "blogOwnerId": "fdda1b48-6a3d-4d94-9156-6118ecc15aff",
                "id": "d4512a1c-ee69-4610-81a0-c00c637294a8",
                "name": "The Blog",
              },
            ],
            "$types": {
              "$": {
                "1._deleted": "undef",
                "1._lastChangedAt": "undef",
                "1._version": "undef",
              },
            },
          },
          {
            "$": [
              5,
              {
                "_deleted": null,
                "_lastChangedAt": null,
                "_version": null,
                "id": "d5e7eb84-2aa1-4bb2-8e2a-21ccc06d2d69",
                "name": "Avatar: Last Airbender",
                "owner": {
                  "_deleted": null,
                  "_lastChangedAt": null,
                  "_version": null,
                  "id": "8619a3a0-2c6b-4e02-8cdb-db3b7f938d4a",
                  "name": "Owner 1",
                },
              },
            ],
            "$types": {
              "$": {
                "1._deleted": "undef",
                "1._lastChangedAt": "undef",
                "1._version": "undef",
                "1.owner._deleted": "undef",
                "1.owner._lastChangedAt": "undef",
                "1.owner._version": "undef",
              },
            },
          },
          {
            "$": [
              6,
              {
                "_deleted": null,
                "_lastChangedAt": null,
                "_version": null,
                "blogOwnerId": "0abe1eec-022b-41aa-9d13-46dccd42ce32",
                "id": "dd83c7af-c316-4e75-8725-f64ffe3055c5",
                "name": "blog2",
              },
            ],
            "$types": {
              "$": {
                "1._deleted": "undef",
                "1._lastChangedAt": "undef",
                "1._version": "undef",
              },
            },
          },
        ],
        "tableName": "user_Blog",
      },
    ],
    "databaseName": "amplify-datastore",
    "databaseVersion": 0.3,
    "tables": [
      {
        "name": "datastore_Setting",
        "rowCount": 1,
        "schema": "++,&[id]",
      },
      {
        "name": "sync_ModelMetadata",
        "rowCount": 0,
        "schema": "++,&[id]",
      },
      {
        "name": "sync_MutationEvent",
        "rowCount": 0,
        "schema": "++,&[id]",
      },
      {
        "name": "user_Blog",
        "rowCount": 6,
        "schema": "++,&[id]",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;
